File: app/src/main/java/com/yineng/yishizhizun/base/BaseWebViewActivity.java
package com.yineng.yishizhizun.base;

import android.graphics.PixelFormat;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;

import com.yineng.yishizhizun.R;
import com.yineng.yishizhizun.callback.WebViewLoadProgressCallback;
import com.yineng.yishizhizun.common.WebViewManager;
import com.yineng.yishizhizun.common.js.JSEnv;
import com.yineng.yishizhizun.util.Logger;
import com.yineng.yishizhizun.widget.AndroidWebView;
import com.yineng.yishizhizun.widget.X5Chrome;

/**
 * Created by "iqos_jay@outlook.com" on 2018/6/22.
 * BaseWebViewActivity
 */
public abstract class BaseWebViewActivity extends BaseActivity implements WebViewLoadProgressCallback {
    private AndroidWebView mAndroidWebView/*Google Android WebView*/;
    protected X5Chrome mX5Chrome/*X5 Chrome*/;
    protected View mWebView;
    public static boolean QRCodeEnable;//二维码扫描
    public static int JSQRCodeScanCallBackID = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Window window = getWindow();
        //设置主界面背景颜色
        Drawable drawable = getResources().getDrawable(R.drawable.main_app_background);
        window.setBackgroundDrawable(drawable);
        window.setFormat(PixelFormat.TRANSLUCENT);
        // 避免输入法界面弹出后遮挡输入光标的问题
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE | WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN);
    }

    /**
     * Create a new WebView to show data
     */
    protected final void createWebView() {
        boolean x5Core = WebViewManager.isX5Core();
        JSEnv.setEnv(JSEnv.CONTEXT, this);
        JSEnv.setEnv(JSEnv.ACTIVITY, this);
        if (!x5Core) {
            mAndroidWebView = new AndroidWebView(this);
            mAndroidWebView.setLoadCallback(this);
            mWebView = mAndroidWebView;
        } else {
            mX5Chrome = new X5Chrome(this);
            mX5Chrome.setLoadCallback(this);
            mWebView = mX5Chrome;
        }
        Logger.verbose("Base", x5Core ? "x5" : "系统");
    }


    /**
     * Activity生命周期->销毁
     */
    @Override
    protected void onDestroy() {
        if (null != mX5Chrome)
            mX5Chrome.destroy();
        if (null != mAndroidWebView)
            mAndroidWebView.destroy();
        super.onDestroy();
    }
}

File: app/src/main/java/com/kupay/kupay/base/BaseWebViewActivity.kt
package com.kupay.kupay.base

import android.bluetooth.le.AdvertisingSetCallback
import android.graphics.PixelFormat
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import com.kupay.kupay.R
import com.kupay.kupay.callback.WebViewLoadProgressCallback
import com.kupay.kupay.common.WebViewManager
import com.kupay.kupay.common.js.JSBridge
import com.kupay.kupay.common.js.JSEnv
import com.kupay.kupay.common.js.JSIntercept
import com.kupay.kupay.util.Logger
import com.kupay.kupay.widget.AndroidWebView
import com.kupay.kupay.widget.X5Chrome
import java.util.HashMap

/**
 * Created by "iqos_jay@outlook.com" on 2018/6/22.
 * BaseWebViewActivity
 */
abstract class BaseWebViewActivity : BaseActivity(), WebViewLoadProgressCallback {
    private var mAndroidWebView/*Google Android WebView*/: AndroidWebView? = null
    protected var mX5Chrome/*X5 Chrome*/: X5Chrome? = null
    protected lateinit var mWebView: View
    private var inited = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val window = window
        //设置主界面背景颜色
        val drawable = resources.getDrawable(R.drawable.main_app_background)
        window.setBackgroundDrawable(drawable)
        window.setFormat(PixelFormat.TRANSLUCENT)
        // 避免输入法界面弹出后遮挡输入光标的问题
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE or WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN)
    }

    /**
     * Create a new WebView to show data
     */
    protected fun createWebView() {
        val x5Core = WebViewManager.isX5Core()
        if (!x5Core) {
            mAndroidWebView = AndroidWebView(this)
            mAndroidWebView!!.setLoadCallback(this)
            mAndroidWebView?.addJavascriptInterface(JSBridge(), JSBridge::class.java.simpleName)
            mAndroidWebView?.addJavascriptInterface(JSIntercept(), JSIntercept::class.java.simpleName)
            JSEnv.setEnv(JSEnv.WEBVIEW, mAndroidWebView)
            mWebView = mAndroidWebView as AndroidWebView
        } else {
            mX5Chrome = X5Chrome(this)
            mX5Chrome!!.setLoadCallback(this)
            //设置要放到这里来写、因为添加的时候是全局静态的HashMap
            //放到WebView中会被已经Destroy了的WebView覆盖掉
            mX5Chrome?.addJavascriptInterface(JSBridge(), JSBridge::class.java.simpleName)
            mX5Chrome?.addJavascriptInterface(JSIntercept(), JSIntercept::class.java.simpleName)
            JSEnv.setEnv(JSEnv.WEBVIEW, mX5Chrome)
            mWebView = mX5Chrome as X5Chrome
        }
        JSEnv.setEnv(JSEnv.CONTEXT, this)
        JSEnv.setEnv(JSEnv.ACTIVITY, this)
        inited = true
        Logger.verbose("Base", if (x5Core) "x5" else "系统")
    }

    /**
     * 加载URL
     */
    protected fun loadUrlIntoWebView(url: String) {
        val extraHeaders = HashMap<String, String>()
        extraHeaders["Referer"] = url
        if (WebViewManager.isX5Core()) mX5Chrome?.loadUrl(url, extraHeaders)
        else mAndroidWebView?.loadUrl(url, extraHeaders)
    }

    override fun onResume() {
        super.onResume()
        if (!inited) return
        if (null != mAndroidWebView) JSEnv.setEnv(JSEnv.WEBVIEW, mAndroidWebView)
        else if (null != mX5Chrome) JSEnv.setEnv(JSEnv.WEBVIEW, mX5Chrome)
    }

    /**
     * Activity生命周期->销毁
     */
    override fun onDestroy() {
        mX5Chrome?.destroy()
        mAndroidWebView?.destroy()
        super.onDestroy()
    }

}
